glyphs-reader/src/propagate_anchors.rs

```diff
@@ -0,0 +1,883 @@
+    // then only update the actual glyphs after we've done all the work.
+    let mut all_anchors = HashMap::new();
+    for name in todo {
+        // temporarily remove the glyph so we can mess with it without borrowing all glyphs
+        let glyph = glyphs.get(&name).unwrap();
+        // we work with indices here to also get around borrowck, because
+        // if we borrow the layers in the loop we can't set the anchors in the loop body
+        for layer in &glyph.layers {
+            let anchors = anchors_traversing_components(
+                glyph,
+                layer,
+                glyphs,
+                &all_anchors,
+                &mut num_base_glyphs,
+            );
+            if anchors.len() != layer.anchors.len() {
+                debug_new_anchors(&anchors, glyph, layer);
+            }
+            all_anchors.entry(name.clone()).or_default().push(anchors);
+        }
+    }
+
+    // finally update our glyphs with the new anchors, where appropriate
+    for (name, layers) in all_anchors {
+        let glyph = glyphs.get_mut(&name).unwrap();
+        if glyph.has_components() {
+            assert_eq!(layers.len(), glyph.layers.len());
+            for (i, layer_anchors) in layers.into_iter().enumerate() {
+                glyph.layers[i].anchors = layer_anchors;
+            }
+        }
+    }
+}
+
+fn debug_new_anchors(anchors: &[Anchor], glyph: &Glyph, layer: &Layer) {
+    if !log::log_enabled!(log::Level::Info) {
+        return;
+    }
+    let prev_names: Vec<_> = layer.anchors.iter().map(|a| &a.name).collect();
+    let new_names: Vec<_> = anchors.iter().map(|a| &a.name).collect();
+    log::info!(
+        "propagated anchors for ('{}': {prev_names:?} -> {new_names:?}",
+        glyph.name,
+    );
+}
+
+/// Return the anchors for this glyph, including anchors from components
+fn anchors_traversing_components<'a>(
+    glyph: &'a Glyph,
+    layer: &'a Layer,
+    glyphs: &BTreeMap<SmolStr, Glyph>,
+    // map of glyph -> anchors, per-layer, updated as we do each glyph
+    visited_anchors: &HashMap<SmolStr, Vec<Vec<Anchor>>>,
+    // each (glyph, layer) writes its number of base glyphs into this map during traversal
+    base_glyph_counts: &mut HashMap<(SmolStr, &'a str), usize>,
+) -> Vec<Anchor> {
+    if layer.anchors.is_empty() && layer.components().count() == 0 {
+        return Vec::new();
+    }
+
+    // if this is a mark and it has anchors, just return them.
+    if !layer.anchors.is_empty() && glyph.category == Some(Category::Mark) {
+        let mut anchors = layer.anchors.clone();
+        // if there is an '*origin' anchor, other positions are adjusted by its pos:
+        if let Some(origin) = anchors
+            .iter()
+            .position(Anchor::is_origin)
+            .map(|idx| anchors.remove(idx))
+        {
+            anchors
+                .iter_mut()
+                .for_each(|a| a.pos -= origin.pos.to_vec2())
+        }
+        return anchors;
```

Comment:
Suggest extracting this into it's own function

Reply:
nice we can also reuse that in the other place we handle "*origin", below.